### PR TITLE
Bulk load CDK: add truncate refresh test

### PR DESCRIPTION
**Pull request by edgao**
* add a test for truncate refresh
* update the mock destination backend to support truncating

### CODE REVIEW

**Comment by edgao** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/46955?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#46955** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/46955?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#46954** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/46954?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#46749** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/46749?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#46922** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/46922?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#46921** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/46921?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#46920** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/46920?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#46950** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/46950?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @edgao and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-cdk/bulk/core/load/src/integrationTest/kotlin/io/airbyte/cdk/load/mock_integration_test/MockDestinationWriter.kt [hunk 1]
```diff
@@ -19,6 +20,10 @@ import javax.inject.Singleton
 @Singleton
 class MockDestinationWriter : DestinationWriter {
     override fun createStreamLoader(stream: DestinationStream): StreamLoader {
+        MockDestinationBackend.retainNewRecords(
```
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/integrationTest/kotlin/io/airbyte/cdk/load/mock_integration_test/MockDestinationWriter.kt:23]: I think this is idiomatic, but lmk if I'm supposed to emitting a batch or something

File: airbyte-cdk/bulk/core/load/src/integrationTest/kotlin/io/airbyte/cdk/load/mock_integration_test/MockDestinationWriter.kt [hunk 2]
```diff
@@ -19,6 +20,10 @@ import javax.inject.Singleton
 @Singleton
 class MockDestinationWriter : DestinationWriter {
     override fun createStreamLoader(stream: DestinationStream): StreamLoader {
+        MockDestinationBackend.deleteOldRecords(
```
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/integrationTest/kotlin/io/airbyte/cdk/load/mock_integration_test/MockDestinationWriter.kt:23]: I think this is idiomatic, but lmk if I'm supposed to emitting a batch or something

File: airbyte-cdk/bulk/core/load/src/integrationTest/kotlin/io/airbyte/cdk/load/mock_integration_test/MockDestinationBackend.kt [hunk 3]
```diff
@@ -21,6 +21,12 @@ object MockDestinationBackend {
         return getFile(filename)
     }
 
+    fun deleteOldRecords(filename: String, minGenerationId: Long) {
```
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/integrationTest/kotlin/io/airbyte/cdk/load/mock_integration_test/MockDestinationBackend.kt:24]: do you think we can/should have a MockObjectStorage integration test? i.e. something to exercise your new toolkit, without actually relying on s3

or is s3 fast enough that that's not really necessary?

File: airbyte-cdk/bulk/core/load/src/integrationTest/kotlin/io/airbyte/cdk/load/mock_integration_test/MockDestinationWriter.kt [hunk 4]
```diff
@@ -31,6 +32,13 @@ class MockStreamLoader(override val stream: DestinationStream) : StreamLoader {
         override val state = Batch.State.PERSISTED
     }
 
+    override suspend fun start() {
+        MockDestinationBackend.deleteOldRecords(
```
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/integrationTest/kotlin/io/airbyte/cdk/load/mock_integration_test/MockDestinationWriter.kt:36]: I think this is idiomatic, but lmk if it belongs somewhere else

File: airbyte-cdk/bulk/core/load/src/integrationTest/kotlin/io/airbyte/cdk/load/mock_integration_test/MockDestinationBackend.kt (hunk 3 above)
Comment (johnny-schmidt) [reply airbyte-cdk/bulk/core/load/src/integrationTest/kotlin/io/airbyte/cdk/load/mock_integration_test/MockDestinationBackend.kt:24]: Unit tests around the s3 interface, with an actual s3 for validation, makes more sense? Then separate unit tests around avro, csv, parquet conversion, using a mocked interface that dumps to a local file.

**Review by johnny-schmidt — APPROVED** [review]

**Pull request merged**